pfrl/agents/option_critic.py

```diff
@@ -0,0 +1,239 @@
+# Adapted from https://github.com/lweitkamp/option-critic-pytorch
+
+import torch
+import torch.nn as nn
+import numpy as np
+import random
+from math import exp
+from torch.distributions import Categorical, Bernoulli
+from pfrl import agent
+from pfrl.replay_buffers.hrl_replay_buffer import OptionCriticReplayBuffer
+from pfrl.replay_buffer import batch_experiences_with_option, ReplayUpdater
+from copy import deepcopy
+
+class OptionCriticNetwork(nn.Module):
+    """Option Critic Network
+
+    This class is used to instantiate the Option Critic Architecture.
+
+    Args:
+        featureNetwork (torch.nn.Module): Network to convert from state to features
+        terminationNetwork (torch.nn.Module): Network to determine termination probs
+        QNetwork (torch.nn.Module): Network to calculate Q-Value
+        feature_output_size (int): size of feature vector outputted by featureNetwork
+        num_options (int): Number of options used
+        num_actions (int): Number of actions available
+        device (string): device to put network on
+        eps_start (float): starting epsilon
+        eps_min (float): minimum epsilon
+        eps_decay (int): decay rate for epsilon
+
+    """
+    def __init__(self, featureNetwork, terminationNetwork, QNetwork, feature_output_size, num_options, num_actions, device='cpu',
+                 eps_start = 1.0, eps_min = 0.1, eps_decay = int(1e6)):
+        super(OptionCriticNetwork, self).__init__()
+
+        self.eps_start = eps_start
+        self.eps_min = eps_min
+        self.eps_decay = eps_decay
+        self.num_steps = 0
+
+        self.features = featureNetwork
+        self.terminations = terminationNetwork
+        self.Q = QNetwork
+        self.options_W = nn.Parameter(torch.zeros(num_options, feature_output_size, num_actions))
+        self.options_b = nn.Parameter(torch.zeros(num_options, num_actions))
+
+        self.device = device
+
+        self.to(device)
+
+    def get_state(self, obs):
+        """Convert state to features."""
+        obs = obs.to(self.device)
+        state = self.features(obs)
+        return state
+
+    def get_Q(self, state):
+        """Get Q value from state."""
+        return self.Q(state)
+
+    def predict_option_termination(self, state, current_option):
+        """Predict whether option will terminate, return (termination, next option)."""
+        termination = self.terminations(state)[current_option].sigmoid()
+        option_termination = Bernoulli(termination).sample()
+
+        return bool(option_termination.item())
```

Comment:
Was this mentioned somewhere in the paper? I just want to know where this is coming from.

Reply:
The use of sigmoid functions on the end of the termination functions is in the experiments section of the paper. Each termination function returns a probability of terminating given a state and current option. We then use the Bernoulli distribution to pick whether it ends or not using the probability generated.